adafruit_emc2101/emc2101_fanspeed.py

```diff
@@ -0,0 +1,201 @@
+# SPDX-FileCopyrightText: Copyright (c) 2020 Bryan Siepert for Adafruit Industries
+#
+# SPDX-License-Identifier: MIT
+"""
+`adafruit_emc2101.emc2101_fanspeed`
+================================================================================
+
+Brushless fan controller: extended functionality
+
+
+* Author(s): Bryan Siepert, Ryan Pavlik
+
+Implementation Notes
+--------------------
+
+**Hardware:**
+
+* `Adafruit EMC2101 Breakout <https://adafruit.com/product/4808>`_ (Product ID: 4808)
+
+**Software and Dependencies:**
+
+* Adafruit CircuitPython firmware for the supported boards:
+  https://circuitpython.org/downloads
+
+* Adafruit's Bus Device library:
+  https://github.com/adafruit/Adafruit_CircuitPython_BusDevice
+
+* Adafruit's Register library:
+  https://github.com/adafruit/Adafruit_CircuitPython_Register
+"""
+
+from adafruit_register.i2c_struct_array import StructArray
+
+from adafruit_emc2101 import emc2101_regs
+
+__version__ = "0.0.0-auto.0"
+__repo__ = "https://github.com/adafruit/Adafruit_CircuitPython_EMC2101.git"
+
+
+class FanSpeedLUT:
+    """A class used to provide a dict-like interface to the EMC2101's
+    Temperature to Fan speed Look Up Table (LUT).
+
+    Keys are integer temperatures, values are fan duty cycles between 0 and
+    100. A max of 8 values may be stored.
+
+    To remove a single stored point in the LUT, assign it as `None`.
+
+    Usage:
+    At 50 C the speed should be 62% of max duty cycle. (Updates the chip
+    immediately).
+
+    .. code-block:: python
+
+        emc2101.lut[50] = 62
+
+    Set two values up and write to the chip once.
+
+    .. code-block:: python
+
+        with emc2101.lut as lut:
+            lut[20] = 0
+            lut[40] = 10
+
+    Delete an unneeded temperature point: (Updates the chip immediately).
+
+    .. code-block:: python
+
+        emc2101.lut[20] = None
+
+    Read a dict of the currently set values:
+
+    .. code-block:: python
+
+        values = emc2101.lut.lookup_table
+        # returns:
+        # { 40: 10, 50: 62 }
+
+    Delete all current LUT values:
+
+    .. code-block:: python
+
+        for temp in emc2101.lut.lookup_table:
+            emc2101.lut[temp] = None
+    """
+
+    # 8 (Temperature, Speed) pairs in increasing order
+    _fan_lut = StructArray(emc2101_regs.LUT_BASE, "<B", 16)
+    _defer_update = 0
+
+    def __init__(self, fan_obj):
+        self._defer_update = 0
+        self.emc_fan = fan_obj
+        self.lut_values = {}
+        self.i2c_device = fan_obj.i2c_device
+
+    def __enter__(self):
+        """'with' wrapper: defer lut update until end of 'with' so
+        update_lut work can be done just once at the end of setting the LUT.
+
+        """
+        # Use increment/decrement so nested with's are dealt with properly.
+        self._defer_update += 1
```

Comment:
I think the use of `_defer_update` to account for nested uses of `with` isn't needed since there shouldn't ever be a case where nested `with` for the same device should be done.  This is also true for `__exit__()`.  I also think it goes against the concept of the context manager if it doesn't perform the intended exit action upon leaving, so I would strongly advise on changing that behavior.

In essence, if someone is using nested `with` statements, it's probably more likely their code isn't working as they intended anyway, and it's better to fail so that can be addressed, rather than try to fix things in the background and mask other problems.

Reply:
I think regarding the use of allowing `__setitem__()` to work within and outside of a context manager, a `bool` internal variable would be more appropriate since it could hold the state only long as a the context manager is being used.

Reply:
This change has been made.